numba/core/bytecode.py

```diff
@@ -321,45 +361,199 @@ def get_used_globals(self):
 
 
 class ByteCodePy311(_ByteCode):
+
     def __init__(self, func_id):
         super().__init__(func_id)
-
-        def fixup_eh(ent):
-            from dis import _ExceptionTableEntry
-            # Patch up the exception table offset
-            # because we add a NOP in _patched_opargs
-            out = _ExceptionTableEntry(
-                start=ent.start + _FIXED_OFFSET, end=ent.end + _FIXED_OFFSET,
-                target=ent.target + _FIXED_OFFSET,
-                depth=ent.depth, lasti=ent.lasti,
-            )
-            return out
-
         entries = dis.Bytecode(func_id.code).exception_entries
-        self.exception_entries = tuple(map(fixup_eh, entries))
+        self.exception_entries = tuple(map(self.fixup_eh, entries))
+
+    @staticmethod
+    def fixup_eh(ent):
+        # Patch up the exception table offset
+        # because we add a NOP in _patched_opargs
+        out = dis._ExceptionTableEntry(
+            start=ent.start + _FIXED_OFFSET, end=ent.end + _FIXED_OFFSET,
+            target=ent.target + _FIXED_OFFSET,
+            depth=ent.depth, lasti=ent.lasti,
+        )
+        return out
 
     def find_exception_entry(self, offset):
         """
         Returns the exception entry for the given instruction offset
         """
         candidates = []
         for ent in self.exception_entries:
-            if ent.start <= offset <= ent.end:
+            if ent.start <= offset < ent.end:
                 candidates.append((ent.depth, ent))
         if candidates:
             ent = max(candidates)[1]
             return ent
 
 
+class ByteCodePy312(ByteCodePy311):
+
+    def __init__(self, func_id):
+        super().__init__(func_id)
+
+        # initialize lazy property
+        self._ordered_offsets = None
+
+        # Fixup offsets for all exception entries.
+        entries = [self.fixup_eh(e) for e in
+                   dis.Bytecode(func_id.code).exception_entries
+                   ]
+
+        # Remove exceptions, innermost ones first
+        # Can be done by using a stack
+        entries = self.remove_build_list_swap_pattern(entries)
+
+        # If this is a generator, we need to skip any exception table entries
+        # that point to the exception handler with the highest offset.
+        if func_id.is_generator:
+            # Get the exception handler with the highest offset.
+            max_exception_target = max([e.target for e in entries])
+            # Remove any exception table entries that point to that exception
+            # handler.
+            entries = [e for e in entries if e.target != max_exception_target]
+            #breakpoint()
+
+        self.exception_entries = tuple(entries)
+
+    @property
+    def ordered_offsets(self):
+        if not self._ordered_offsets:
+            # Get an ordered list of offsets.
+            self._ordered_offsets = [o for o in self.table]
+        return self._ordered_offsets
+
+    def remove_build_list_swap_pattern(self, entries):
```

Comment:
Do we have a testcase for this? if not, i used the following to exercise the nested list-comp case:

```python
@njit
def foo(n):
    return [[i] + [j for j in range(sum([x for x in range(i)]))] for i in range(n)]

r = foo(10)
```

Reply:
I'm reasonably sure the various comprehension tests cover this, e.g. https://github.com/numba/numba/blob/a4664180ddc91e4c8a37dd06324f4e229a76df91/numba/tests/test_comprehension.py#L314-L321

However, it would be good to add a couple of explicit tests for the most problematic cases, which were mostly to do with comprehensions inside try-blocks.